# Data cleaning and exploration of the Ebay Kleinanzeigen dataset

The aim of this project is to clean the Ebay Kleinanzeigen dataset and investigate trends in the car listings. This will include analysing trends in listings and characteristics of popular car listings. This analysis should provide insight into market trends, as well as provide approximate expectations for sellers in this market.

The dataset for this project can be found [here](https://www.kaggle.com/orgesleka/used-cars-database/data)

In [1]:
import pandas as pd
import numpy as np

In [174]:
autos = pd.read_csv("autos.csv", encoding = 'Latin-1')

In [175]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [176]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

As seen above, a majority of the columns appear to be populated with values with the exception of notRepairedDamage, fuelType, model, gearbox and vehicleType. From the columns with missing data, less than 20% of the data is missing. It should also be noted that the majority of columns are strings. The naming conventions are also found to be camelcase rather than snakecase and thus will require cleaning.

## Cleaning the data

Investigate column names

In [177]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [178]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']

In [179]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Changes to columns:
* Converted columns from camelcase to snakecase
* Renamed unclear columns to more intuitive column names

## Exploring dataset 

Display descriptive statistics

In [180]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

From the statistics presented above the following observations were made:
* Columns seller and offer_type are comprised of almost identical values across rows
* num_pictures appears strange and will require further investigation

In [181]:
# investigate num_pictures
autos['num_pictures'].unique()

array([0])

The results presented above indicates that all row values for this column are zero and thus provide no useful insight into analysing the dataset further. This is also true for the offer_type and seller columns which both contain 50,000 of the same values across the column rows.

## Cleaning column data 

From the results presented above, it appears that both the price and odometer columns are numeric values stored as strings. Therefore, to analyse the dataset further, these two columns will be investigated and values converted to numeric.

In [182]:
# investigate price
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

The above results indicate that the price of a car is stored as a string with a \$ at the front and uses comma notation. To clean this column, the \$ will be stripped from the row values, the comma removed and the values casted as integers

In [183]:
# removing $
autos['price'] = pd.to_numeric(autos['price'].str.replace('$','').str.replace(',',''))

In [184]:
# checking results
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

As seen above, the values in the column price have now been converted into integer values without any special characters

In [185]:
# investigating odometer_km
autos['odometer'].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

From the investigation above it can be seen that the values in this column use comma notation and are preceeded by 'km'. To clean this column, the comma will be removed and the non numeric values stripped.

In [186]:
# removing comma notation
autos['odometer'] = autos['odometer'].str.replace(',','').str.replace('km','').astype(int)
autos.rename({'odometer':'odometer_km'}, axis = 1, inplace=True)

In [187]:
# checking results
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

As seen above, the values in this column have now been cleaned and cast as integer values.

## Further exploration of price and odometer

#### Exploring the price column

In [188]:
# unique elements
autos['price'].unique().shape

(2357,)

In [189]:
# descriptive statistics for the column
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [190]:
# view the top 20 highest prices
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [191]:
# view the lowest 20 prices
autos['price'].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

Investigating the prices columns reveals that there are 1421 listings with a price of \$ 0. This is likely an error and thus will be removed. Though logically unlikely, the probability of a \$ 1 is not impossible and thus values equal to or greater than \$ 1 will be kept. 

With regards to the top 20 prices, the assumption that a car sold on eBay would not exceed \$ 350,000 is made. Although this value is high already, the difference between 350k and 990k is too large to be considered reasonable.

In [192]:
# remove rows where the row value of price is not between 1 and 3510000
autos = autos[autos['price'].between(1,351000)]

In [193]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count   48565.000000       48565.000000  48565.000000   48565.000000   
mean     5888.935591        2004.755421    117.197158  125770.101925   
std      9059.854754          88.643887    200.649618   39788.636804   
min         1.000000        1000.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7490.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month  num_pictures   postal_code  
count        48565.000000       48565.0  48565.000000  
mean             5.782251           0.0  50975.745207  
std              3.685595           0.0  25746.968398  
min              0.000000           0.0   1067.000000  
25%              3.000000           0.0  30657.000000  
50%              6.000000           0.0  49716.000000  
75%              9.000000           0.0  71665.000000  
max             12.000000           0.0  99998.000000

As seen above, the dataset has been reduced from 50,000 rows to 48,565 rows as a result of filtering the price values.

#### Exploring the odometer column


In [194]:
# Investigate unique elements
autos['odometer_km'].unique().shape

(13,)

In [195]:
# Investigate descriptive statistics for the column
autos['odometer_km'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [196]:
# Investigate the values
autos['odometer_km'].value_counts()

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64

The values in this column appear to be rounded and thus indicates that the sellers had to choose from pre-set options in this field.

## Investigating the date column 

These following columns contain date information and should be investigated further for potential inconsistencies:
* date_crawled
* last_seen
* ad_created
* registration_month
* registration_year

The values in date_crawled, ad_created and last_seen are strings representing date timestamps. Therefore, these column values will be explored further and converted into proper dates.

In [197]:
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [198]:
# date_crawled
autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05 14:06:30    0.000021
2016-03-05 14:06:40    0.000021
2016-03-05 14:07:04    0.000021
2016-03-05 14:07:08    0.000021
2016-03-05 14:07:21    0.000021
2016-03-05 14:07:26    0.000021
2016-03-05 14:07:40    0.000021
2016-03-05 14:07:45    0.000021
2016-03-05 14:08:00    0.000041
2016-03-05 14:08:05    0.000041
2016-03-05 14:08:27    0.000021
2016-03-05 14:08:42    0.000021
2016-03-05 14:09:02    0.000041
2016-03-05 14:09:05    0.000021
2016-03-05 14:09:20    0.000021
2016-03-05 14:09:22    0.000021
2016-03-05 14:09:38    0.000021
2016-03-05 14:09:46    0.000021
2016-03-05 14:09:56    0.000021
2016-03-05 14:09:57    0.000021
2016-03-05 14:09:58    0.000041
2016-03-05 14:10:18    0.000021
2016-03-05 14:10:20    0.000021
2016-03-05 14:10:46    0.000021
2016-03-05 14:11:03    0.000021
2016-03-05 14:11:05    0.000021
2016-03-05 14:11:14    0.000021
2016-03-05 14:11:15    0.000021
2016-03-05 14:11:25    0.000021
2016-03-05 14:11:40    0.000021
                         ...   
2016-04-

In [199]:
autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_values()

2016-03-14 10:53:11    0.000021
2016-03-12 10:48:42    0.000021
2016-03-24 17:53:16    0.000021
2016-04-01 22:37:14    0.000021
2016-03-16 20:55:47    0.000021
2016-03-31 19:56:44    0.000021
2016-04-04 13:52:58    0.000021
2016-03-12 21:48:13    0.000021
2016-03-30 00:46:40    0.000021
2016-03-28 15:59:52    0.000021
2016-03-29 11:54:16    0.000021
2016-03-18 10:37:18    0.000021
2016-03-19 09:06:24    0.000021
2016-03-22 20:44:27    0.000021
2016-03-08 22:38:50    0.000021
2016-03-07 12:48:43    0.000021
2016-03-05 14:56:51    0.000021
2016-03-19 18:43:22    0.000021
2016-03-30 16:56:27    0.000021
2016-03-14 13:57:33    0.000021
2016-03-26 20:47:12    0.000021
2016-03-19 09:56:17    0.000021
2016-03-07 19:41:45    0.000021
2016-04-01 21:49:40    0.000021
2016-03-28 20:44:40    0.000021
2016-03-15 18:57:08    0.000021
2016-03-09 20:52:08    0.000021
2016-03-22 20:53:22    0.000021
2016-03-11 02:59:24    0.000021
2016-04-05 11:57:47    0.000021
                         ...   
2016-04-

The results above indicate that the site was crawled daily during March and April of 2016. The distribution of this column is close to uniform.

In [200]:
# last_seen
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [201]:
autos['last_seen'].str[0:10].value_counts(normalize=True, dropna=False).sort_values()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-18    0.007351
2016-03-08    0.007413
2016-03-13    0.008895
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-14    0.012602
2016-03-27    0.015649
2016-03-19    0.015834
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-26    0.016802
2016-03-23    0.018532
2016-03-25    0.019211
2016-03-24    0.019767
2016-03-21    0.020632
2016-03-20    0.020653
2016-03-28    0.020859
2016-03-22    0.021373
2016-03-29    0.022341
2016-04-01    0.022794
2016-03-31    0.023783
2016-03-12    0.023783
2016-04-04    0.024483
2016-03-30    0.024771
2016-04-02    0.024915
2016-04-03    0.025203
2016-03-17    0.028086
2016-04-05    0.124761
2016-04-07    0.131947
2016-04-06    0.221806
Name: last_seen, dtype: float64

The above results display the number of last_seens per day. Therefore, using this column we can determine when a listing was removed/sold due to a change in the row value.

However, the results also indicate that there is a large disproportionality of last_seens in the last few dates. This may represent anomolous data or views being carried over across months.

In [202]:
# ad_created
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

In [203]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_values()

2016-01-07    0.000021
2016-01-22    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2016-02-16    0.000021
2016-01-13    0.000021
2016-02-11    0.000021
2016-02-17    0.000021
2015-06-11    0.000021
2016-02-08    0.000021
2016-01-16    0.000021
2015-12-30    0.000021
2015-08-10    0.000021
2015-12-05    0.000021
2016-01-29    0.000021
2016-02-01    0.000021
2016-01-14    0.000021
2016-02-07    0.000021
2016-02-22    0.000021
2016-02-09    0.000021
2016-01-03    0.000021
2016-02-26    0.000041
2016-02-20    0.000041
2016-02-18    0.000041
2016-02-02    0.000041
2016-02-14    0.000041
2016-02-05    0.000041
2016-02-12    0.000041
2016-02-24    0.000041
2016-01-10    0.000041
                ...   
2016-03-06    0.015320
2016-03-13    0.017008
2016-03-05    0.022897
2016-03-24    0.029280
2016-03-16    0.030125
2016-03-27    0.030989
2016-03-17    0.031278
2016-03-25    0.031751
2016-03-31    0.031875
2016-03-10    0.031895
2016-03-23    0.032060
2016-03-26    0.032266
2016-03-22 

The results above indicate that ads created within a 9 month period of the last date were included.

#### Investigating the registration_year column 

In [204]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Exploration of the column revleas that dates recorded range from the year 1000 to the year 9999. This is evidently an error and thus this column should be cleaned to contain only rows with valid registration year dates.

Due to the dataset being created in 2016, the maximum date allowed will be assumed as 2016 whilst the oldest registration year will be assumed to be the year 1900.

Determine percentage of data that has a date outside the range of 1900 - 2016:

In [205]:
within_daterange = autos['registration_year'].between(1900, 2016).value_counts()

In [206]:
within = within_daterange[1]
outside = within_daterange[0]

In [207]:
outside

1884

In [208]:
percentage_outside = outside/(within+outside)

In [209]:
percentage_outside

0.038793369710697

As seen above, approximately only 3.9% of the rows in this column have invalid dates. Therefore, these rows should be removed.

Remove invalid row values

In [2]:
autos = autos[autos['registration_year'].between(1900, 2016)]

NameError: name 'autos' is not defined

Check size of dataset

In [3]:
autos.describe()

NameError: name 'autos' is not defined

#### Calculating the distribution of the remaining values

In [212]:
autos['registration_year'].value_counts(normalize=True).sort_index()

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
1938    0.000021
1939    0.000021
1941    0.000043
1943    0.000021
1948    0.000021
1950    0.000064
1951    0.000043
1952    0.000021
1953    0.000021
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000493
1961    0.000129
1962    0.000086
1963    0.000171
1964    0.000257
1965    0.000364
1966    0.000471
1967    0.000557
1968    0.000557
          ...   
1987    0.001542
1988    0.002892
1989    0.003727
1990    0.007433
1991    0.007262
1992    0.007926
1993    0.009104
1994    0.013474
1995    0.026285
1996    0.029412
1997    0.041794
1998    0.050620
1999    0.062060
2000    0.067608
2001    0.056468
2002    0.053255
2003    0.057818
2004    0.057904
2005    0.062895
2006    0.057197
2007    0.048778
2008    0.047450
2009    0.044665
2010    0.034040
2011    0.034768
2012    0.028063
2013    0.017202
2014    0.0142

The results above show that the remaining values in ths registration_year columns fall within the date range spcified above. It also indicates that there is a larger proportion of newer cars with larger values for more recent dates.

## Exploring the brand column

In [213]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

It appears that European cars represent the top brands. Volkswagen is the most popular brand with almost twice as many car sales as the next highest brand.

As certain brands dont have signficant sales, they will be excluded from our analysis. (Analyse the top 5% of total listings).

In [214]:
brand_percentage = autos['brand'].value_counts(normalize=True)
most_common_brands = brand_percentage[brand_percentage > 0.05].index
print(most_common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [215]:
# extracting brands to store in dictionary
average_brand_price = {}
for brand in most_common_brands:
    brand_name = autos[autos['brand']==brand]
    mean_price = brand_name['price'].mean()
    average_brand_price[brand] = int(mean_price)
    
average_brand_price

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}

From the data presented:
* Audi, Mercedes Benz and BMW's are more expensive
* Volkswagens are in the middle in terms of pricing 
* Fords and Opel are the cheapest of the group

#### Exploring mean mileages 

In [216]:
bmp_series = pd.Series(average_brand_price)
print(bmp_series)

audi             9336
bmw              8332
ford             3749
mercedes_benz    8628
opel             2975
volkswagen       5402
dtype: int64


extracting mean mileage from car brands

In [4]:
mean_mileage = {}
for brand in most_common_brands:
    brands = autos[autos['brand']==brand]
    mileage = brands['odometer_km'].mean()
    mean_mileage[brand] = int(mileage)
    

NameError: name 'most_common_brands' is not defined

Constructing dataframe from price dictionary

In [218]:
bmp_series = pd.Series(average_brand_price).sort_values(ascending=False)
price_df = pd.DataFrame(bmp_series, columns=['mean_price'])

Joining the mean mileage data onto the new dataframe

In [219]:
bmm_series = pd.Series(mean_mileage).sort_values(ascending=False)
price_df['mean_mileage'] = bmm_series[0:]

Display the dataframe

In [220]:
price_df

mean_price  mean_mileage
audi                 9336        129157
mercedes_benz        8628        130788
bmw                  8332        132572
volkswagen           5402        128707
ford                 3749        124266
opel                 2975        129310

From the data presented above, there is no heavy correlation between mileage and price of this data set. This is evident when observing the Opel mean price given a mileage of 129310 to audi with a mileage of 129157 or bmw with a higher price and mileage of 132572. This is likely due to the brand themselves and the premium prices associated with them. However, when investigating the mileage and average price of the luxury cars alone (Bmw, audi, merc), there appears to be a correlation between mean mileage and mean price given that audi with the lowest mileage is the most expensive, bmw with the lowest price and highest mileage and merc which is in the middle for both mileage and price. 

**Additional steps for the future**:
* Clean categorical data using German words, translate them to English and map values to English counterpart
* Change format of date
* Investigate most common brand combinations
* Split the odometer_km into groups, use aggregation to investigate if average prices follows any patterns on the mileage
* Investigate how much cheaper cars with damage are compared to non-damaged counterparts